gitbranchstack/main.py

```diff
@@ -285,8 +285,16 @@ def on_conflict(path):
         global ON_CONFLICT
         ON_CONFLICT = on_conflict
         head = rebase(patch, head)
+        message = head.message
         if t == topic or trim_all_subjects or trim_subject:
-            head = head.update(message=trimmed_message(subject, patch.message))
+            message = trimmed_message(subject, patch.message)
+        head = repo.new_commit(
+            message=message,
+            tree=head.tree(),
+            parents=head.parents(),
+            author=head.author,
+            committer=patch.committer, # preserve original author and timestamp
```

Comment:
Nice, I never thought of preserving the commit date, probably because tools like `git rebase` or `git revise` don't.
By updating the commit date, these tools preserve the invariant that every commit has a later commit date than its parents.

`git branchstack` usually maintains that invariant by design, because commits are applied on top of older commits. There is an exception, if you specify dependencies with a weird ordering, like

	(HEAD) [parent] prerequisite commit 2
	[child:+parent] child commit with dependencies
	[parent] prerequisite commit 1

When generating `child`, all parent commits will be added before the child, so the resulting commit graph is no longer monotonic w.r.t. committer date

	child commit with dependencies
	prerequisite commit 2
	prerequisite commit 1

Though this is a weird case, and we might go back to using the same order as in the user's branch.

---

A few lines down is how I avoided getting different commit IDs across repeated runs of `git branchstack`:

```python
    if head.tree() != topic_ref.target.tree() or new_messages != old_messages:
        topic_oid = topic_ref.target.oid
        print(f"Updating {topic_ref.name} ({topic_oid} => {head.oid})")
        topic_ref.update(head, "git-branchstack rewrite")
```

so we skipped rewriting the branch if the final tree + commit messages are unchanged. This will be wrong in edge cases where you move a content hunk from one commit to the other with the same topic.

With your fix to make generated commit IDs deterministic, we can fix this by always updating when the commit ID changed:

```python
    if head.oid != topic_ref.target.oid:
```

Reply:
> By updating the commit date, these tools preserve the invariant that every commit has a later commit date than its parents.

Oh, that's interesting, it never occurred to me that that would be an invariant of those tools but I guess it makes sense.

> git branchstack usually maintains that invariant by design, because commits are applied on top of older commits. There is an exception, if you specify dependencies with a weird ordering

Honestly I wonder if it would be better to just blow up if the branch has this structure. Something seems inherently wrong about a commit claiming to depend on a commit that it's actually an ancestor of.

It also seems like, if there isn't a conflict with the commits being reordered to match the dependencies, the user could easily do the reordering themselves with an interactive rebase before invoking `branchstack`, and if there is a conflict then `branchstack` won't work anyway, right?

> With your fix to make generated commit IDs deterministic, we can fix this by always updating when the commit ID changed:

Cool, I can add that change here.

Reply:
> Honestly I wonder if it would be better to just blow up if the branch has this structure. Something seems inherently wrong about a commit claiming to depend on a commit that it's actually an ancestor of.

Instead of erroring

> It also seems like, if there isn't a conflict with the commits being reordered to match the dependencies, the user could easily do the reordering themselves with an interactive rebase before invoking `branchstack`,

Yeah, reordering commits with revise/rebase is the canonical solution if dependencies are in the wrong order.

I just realized that my example does not actually work like this anymore - instead we create branches with the same commit order as in the source branch.

We used to apply the topics one by one, but that gave hard-to-predict orderings with transitive dependencies.
Using the order from the source branch is much more sane.

Erroring out on wrong dependency order is possible, why not. No strong opinion here.

> and if there is a conflict then `branchstack` won't work anyway, right?

It works good enough for simple conflicts - just like git-revise it lets you edit conflicting files in your $EDITOR:

> If a commit cannot be applied cleanly, `git branchstack` will show topics that would avoid the conflict if added as dependencies. You can either add the missing dependencies, or resolve the conflict in your editor. You can tell Git to remember your conflict resolution by enabling `git rerere` (use `git config rerere.enabled true; git config rerere.autoUpdate true`). 

So it can manage independent branches that have small conflicts. With rerere you just need to resolve conflicts once.